### PR TITLE
Respect value written to DEPTH when using shadow maps.

### PR DESCRIPTION
**Pull request by basicer**
Responding to the comments in #65307 and possibly also closing [Proposal-1942](https://github.com/godotengine/godot-proposals/issues/1942).

This adjusts the position used to sample shadow maps to take the written DEPTH value into account. For object that both cast and receive shadows (the default), not updating the position causes the fragment to shadows itself in most cases, almost always looking horrible. Because of this it's unlikely there are any uses of depth writing shaders that receive shadows in use today.

I agree with @lyuma in that I don't understand why the other changes are necessary.  This patch only updates the position, and only when DEPTH is actually used.

Before, fragments shadow themselves:
image

After, they dont!
image

Tested on Mac OS on Vulcan

### CODE REVIEW

File: drivers/gles3/shaders/scene.glsl [hunk 1]
```diff
@@ -1515,6 +1515,14 @@ void main() {
 #CODE : FRAGMENT
 	}
 
+#ifdef DEPTH_USED
+	vec3 ndc = vec3(screen_uv * 2.0 - 1.0, gl_FragDepth);
+	vec4 view_pos = scene_data.inv_projection_matrix * vec4(ndc, 1.0);
+	// Vertex is in view space, the rays are not parallel. It is not enough to update the z component!
+	vertex = view_pos.xyz / view_pos.w;
+	view = -normalize(vertex);
```
Comment (lyuma) [inline drivers/gles3/shaders/scene.glsl:1523]: I'm curious why it is necessary to update `view`. My understanding is everything at that pixel coordinate is some `depth` along the `view` vector. Changing the depth should not affect the view vector.

Your comment notes that "Vertex is in view space, the rays are not parallel. It is not enough to update the z component!"

While you are correct that `vertex` and `view` are in view space, `depth` and `screen_uv` are in *clip space* / screen space, and by changing the z component in *clip space*, we are moving along the `view` ray, so `normalize(view)` should stay constant regardless of how we manipulate `depth` (except perhaps if we push it behind the camera)

Comment (basicer) [reply drivers/gles3/shaders/scene.glsl:1523]: I'm not sure, I took the original author's word for it.  It seems to work fine without changing view!

**Comment by basicer** [comment]
[main.zip](https://github.com/godotengine/godot/files/15046643/main.zip)

I made minimal project file showing the issue based on https://godotshaders.com/shader/raymarching-with-depth-writting/ to show the issue.

With out the patch, shadows are kinda crazy.  This is especially noticeable when the large ball moves though the left side of the screen.

https://github.com/godotengine/godot/assets/671513/8cd0b265-777c-4635-abc1-c6103d2ee17c

With the patch, the light looks correct.

https://github.com/godotengine/godot/assets/671513/17fa103b-9a1f-4a71-893d-a0aaffb99427

**Comment by basicer** [comment]
In #65307 @clayjohn seem to agree this was better categorized as a bug, and I've attached the reproduction bugs required.  Could this PR be considered for that track?

File: drivers/gles3/shaders/scene.glsl [hunk 2]
```diff
@@ -1515,6 +1515,12 @@ void main() {
 #CODE : FRAGMENT
 	}
 
+#ifdef DEPTH_USED
+	vec3 ndc = vec3(screen_uv * 2.0 - 1.0, gl_FragDepth);
+	vec4 view_pos = scene_data.inv_projection_matrix * vec4(ndc, 1.0);
```
Comment (clayjohn) [inline drivers/gles3/shaders/scene.glsl:1520]:
```suggestion
	vec4 view_pos = inv_projection_matrix * vec4(ndc, 1.0);
```

Should use this directly so it properly handles multiview

https://github.com/godotengine/godot/blob/4a0160241fd0c1e874e297f6b08676cf0761e5e8/drivers/gles3/shaders/scene.glsl#L1422

**Comment by akien-mga** [comment]
Superseded by #91136.

**Pull request closed without merging**